components/aws/sagemaker/common/common_inputs.py

```diff
@@ -0,0 +1,153 @@
+    """Defines the structure of a component input to be used for validation."""
+
+    input_type: Callable
+    description: str
+    required: bool = False
+    choices: Optional[List[object]] = None
+    default: Optional[object] = None
+
+    def to_argparse_mapping(self):
+        """Maps each property to an argparse argument.
+
+        See: https://docs.python.org/3/library/argparse.html#adding-arguments
+        """
+        return {
+            "type": self.input_type,
+            "help": self.description,
+            "required": self.required,
+            "choices": self.choices,
+            "default": self.default,
+        }
+
+
+@dataclass(frozen=True)
+class SageMakerComponentOutputValidator:
+    """Defines the structure of a component output."""
+
+    description: str
+
+
+# The value of an input or output can be arbitrary.
+# This could be replaced with a generic, as well.
+SageMakerIOValue = NewType("SageMakerIOValue", object)
+
+# Allow the component input to represent either the validator or the value itself
+# This saves on having to rewrite the struct for both types.
+# Could replace `object` with T if TypedDict supported generics (see above).
+SageMakerComponentInput = NewType(
+    "SageMakerComponentInput", Union[SageMakerComponentInputValidator, SageMakerIOValue]
+)
+SageMakerComponentOutput = NewType(
+    "SageMakerComponentOutput",
+    Union[SageMakerComponentOutputValidator, SageMakerIOValue],
+)
+
+
+@dataclass(frozen=True)
+class SageMakerComponentBaseInputs:
+    """The base class for all component inputs."""
+
+    pass
+
+
+@dataclass
+class SageMakerComponentBaseOutputs:
+    """A base class for all component outputs."""
+
+    pass
+
+
+@dataclass(frozen=True)
+class SageMakerComponentCommonInputs(SageMakerComponentBaseInputs):
+    """A set of common inputs for all components."""
+
+    region: SageMakerComponentInput
+    endpoint_url: SageMakerComponentInput
+    assume_role: SageMakerComponentInput
+    tags: SageMakerComponentInput
+
+
+COMMON_INPUTS = SageMakerComponentCommonInputs(
+    region=SageMakerComponentInputValidator(
+        input_type=str,
+        required=True,
+        description="The region for the SageMaker resource.",
+    ),
+    endpoint_url=SageMakerComponentInputValidator(
+        input_type=SpecValidators.nullable_string_argument,
+        required=False,
+        description="The URL to use when communicating with the SageMaker service.",
+    ),
+    assume_role=SageMakerComponentInputValidator(
+        input_type=str,
+        required=False,
+        description="The ARN of an IAM role to assume when connecting to SageMaker.",
+    ),
+    tags=SageMakerComponentInputValidator(
+        input_type=SpecValidators.yaml_or_json_dict,
+        required=False,
+        description="An array of key-value pairs, to categorize AWS resources.",
+        default={},
+    ),
+)
+
+
+@dataclass(frozen=True)
+class SpotInstanceInputs(SageMakerComponentBaseInputs):
```

Comment:
SpotInstanceInputs and SPOT_INSTANCE_INPUTS can create confusion, can we name this little differently and follow same convention for COMMON_INPUTS as well